portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -30,127 +30,217 @@
             response.sendRedirect("/contact.html");
             return;
         }
-
-        // Only update maxComments if tempMax was not negative
-        if (tempMax > 0)
-        {
-            maxComments = tempMax;
+        
+        // Get the input from the form.
+        String comment = getParameter(request, "user-comment", null);
+
+        // Return error message if the user did not input any comment.
+        if (comment == null) {
+            response.setContentType("text/html;");
+            response.getWriter().println("Enter a comment before submitting.");
+            return;
         }
-
-        response.sendRedirect("/contact.html");
-        return;
-    }
     
-    // Get the input from the form.
-    String comment = getParameter(request, "user-comment", null);
+        // If the user did submit a comment, add it to the datastore
+        createComment(comment);
+
+        // Increase total of all comments by 1
+        changeAllCommentsTotal(1);
 
-    // Return error message if the user did not input any comment.
-    if (comment == null) {
+        // Respond with a success message
         response.setContentType("text/html;");
-        response.getWriter().println("Enter a comment before submitting.");
-        return;
+        response.getWriter().println("Your comment has been registered. Thank you!");
     }
-    
-    // If the user did submit a comment, add it to the datastore
-    Entity commentEntity = new Entity("Comment");
-    commentEntity.setProperty("content", comment);
 
-    long timestamp = System.currentTimeMillis();
-    commentEntity.setProperty("time", timestamp);
+    /** Returns the desired parameter entered by the user, or null if the user input was invalid. */
+    private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+        // Get the input from the form.
+        String value = request.getParameter(name);
+
+        if (value == null) {
+            return defaultValue;
+        }
+
+        return value;
+    }
 
-    commentEntity.setProperty("thumbsup", 0);
+    // Iterates over a comments query and returns an array of comments
+    private List<Comment> iterateQuery(PreparedQuery results) {
+        List<Comment> comments = new ArrayList<>();
+        Iterator<Entity> iter = results.asIterator();
+        int totalComments = results.countEntities();
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query query = new Query("AllComments");
+        PreparedQuery resultsAll = datastore.prepare(query);
+        Iterator<Entity> allIter = resultsAll.asIterator();
+        Entity allComments = allIter.next();
+        long page = (long) allComments.getProperty("page");
+
+        for (int count = 0; count < (maxComments * page) && count < totalComments; count++) {
+            Entity entity = iter.next();
+            if (count >= (maxComments * (page - 1))) {
```

Comment:
Maybe add a couple comments here, it took a while for me to understand exactly what was happening here.

Reply:
Yeah it does look confusing. Thanks!